File: sample-projects/csharp/src/De.Sots.CellarsAndCentaurs/Domain/Model/Creature.cs
using System.Collections.Generic;
using De.Sots.CellarsAndCentaurs.Application;

namespace De.Sots.CellarsAndCentaurs.Domain.Model;

/// <summary>
/// A creature that roams the cellar. Centaurs, beasts and dragons all share hit points, armor class and speeds.
/// </summary>
public class Creature : Fightable
{
    public CreatureId Id { get; }

    public CreatureType Type { get; set; }

    public ArmorClass? ArmorClass { get; set; }

    public HitPoints? HitPoints { get; set; }

    public IDictionary<SpeedType, Speed> Speeds { get; set; } = new Dictionary<SpeedType, Speed>();

    public Creature(CreatureId id)
        : this(id, CreatureFacade.STANDARD_CREATURE_TYPE)
    {
    }

    public Creature(CreatureId id, CreatureType type)
    {
        Id = id;
        Type = type;
    }

    public virtual int Initiative() => Dice.RollD20();

    public void TakeDamage(int damage)
    {
        HitPoints = HitPoints?.Damaged(damage);
    }

    public Speed? SpeedOf(SpeedType speedType) => Speeds.TryGetValue(speedType, out var speed) ? speed : null;
}

File: sample-projects/swift/Sources/CellarsAndCentaurs/Domain/Model/Creature.swift
import Logging

/// A creature that roams the cellar. Centaurs, beasts and dragons all share hit points, armor class and speeds.
class Creature: Fightable {
    private static let logger = Logger(label: "de.sots.cellarsandcentaurs.domain.model.Creature")

    let id: CreatureId
    var type: CreatureType
    var armorClass: ArmorClass?
    var hitPoints: HitPoints?
    var speeds: [SpeedType: Speed] = [:]

    init(id: CreatureId, type: CreatureType = CreatureFacade.STANDARD_CREATURE_TYPE) {
        self.id = id
        self.type = type
    }

    var walkingSpeed: Speed {
        speeds[.walking] ?? Speed.stationary
    }

    func takeDamage(_ amount: Int) {
        Creature.logger.info("creature \(id.value) takes \(amount) damage")
        hitPoints = hitPoints?.damaged(by: amount)
    }
}
